flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -0,0 +1,298 @@
+
+		this.jobID = jobID;
+		this.allocationID = allocationID;
+		this.jobVertexID = jobVertexID;
+		this.subtaskIndex = subtaskIndex;
+		this.discardExecutor = discardExecutor;
+		this.lock = new Object();
+		this.storedTaskStateByCheckpointID = new TreeMap<>();
+		this.discarded = false;
+		this.localRecoveryConfig = localRecoveryConfig;
+	}
+
+	@Override
+	public void storeLocalState(
+		@Nonnegative long checkpointId,
+		@Nullable TaskStateSnapshot localState) {
+
+		if (localState == null) {
+			localState = NULL_DUMMY;
+		}
+
+		LOG.info("Storing local state for checkpoint {}.", checkpointId);
+		LOG.debug("Local state for checkpoint {} is {}.", checkpointId, localState);
+
+		Map<Long, TaskStateSnapshot> toDiscard = new HashMap<>(MAX_RETAINED_SNAPSHOTS);
+
+		synchronized (lock) {
+			if (discarded) {
+				// we ignore late stores and simply discard the state.
+				toDiscard.put(checkpointId, localState);
+			} else {
+				TaskStateSnapshot previous =
+					storedTaskStateByCheckpointID.put(checkpointId, localState);
+
+				if (previous != null) {
+					toDiscard.put(checkpointId, previous);
+				}
+
+				// remove from history.
+				while (storedTaskStateByCheckpointID.size() > MAX_RETAINED_SNAPSHOTS) {
+					Long removeCheckpointID = storedTaskStateByCheckpointID.firstKey();
+					TaskStateSnapshot snapshot =
+						storedTaskStateByCheckpointID.remove(removeCheckpointID);
+					toDiscard.put(removeCheckpointID, snapshot);
+				}
+			}
+		}
+
+		asyncDiscardLocalStateForCollection(toDiscard.entrySet());
+	}
+
+	@Override
+	@Nullable
+	public TaskStateSnapshot retrieveLocalState(long checkpointID) {
+		synchronized (lock) {
+			TaskStateSnapshot snapshot = storedTaskStateByCheckpointID.get(checkpointID);
+			return snapshot != NULL_DUMMY ? snapshot : null;
+		}
+	}
+
+	@Override
+	@Nonnull
+	public LocalRecoveryConfig getLocalRecoveryConfig() {
+		return localRecoveryConfig;
+	}
+
+	@Override
+	public void confirmCheckpoint(long confirmedCheckpointId) {
+
+		LOG.debug("Received confirmation for checkpoint {}. Starting to prune history.", confirmedCheckpointId);
+
+		final List<Map.Entry<Long, TaskStateSnapshot>> toRemove = new ArrayList<>(MAX_RETAINED_SNAPSHOTS);
+
+		synchronized (lock) {
+
+			Iterator<Map.Entry<Long, TaskStateSnapshot>> entryIterator =
+				storedTaskStateByCheckpointID.entrySet().iterator();
+
+			// remove entries for outdated checkpoints and discard their state.
+			while (entryIterator.hasNext()) {
+
+				Map.Entry<Long, TaskStateSnapshot> snapshotEntry = entryIterator.next();
+				long entryCheckpointId = snapshotEntry.getKey();
+
+				if (entryCheckpointId < confirmedCheckpointId) {
+					toRemove.add(snapshotEntry);
+					entryIterator.remove();
+				} else {
+					// we can stop because the map is sorted.
+					break;
+				}
+			}
+		}
+
+		asyncDiscardLocalStateForCollection(toRemove);
+	}
+
+	/**
+	 * Disposes the state of all local snapshots managed by this object.
+	 */
+	public void dispose() {
+
```

Comment:
I found this is method is only called via `releaseLocalStateForAllocationId()` and `releaseLocalStateForAllocationId()` is only called in `requestSlot()` and `freeSlotInternal()`, this seem to lead to resource leak. I.e: when user cancel a job normally.

Reply:
If the user cancels the job, then the allocated slot should be released, thus, also releasing the local state data.